File: algorithms/julia/montecarlo.jl
function lcg_next(s::UInt32)
    s = s * UInt32(1664525) + UInt32(1013904223)
    return Float64(s >> 1) / Float64(UInt32(1) << 31), s
end

function montecarlo(n::Int, seed::UInt32)
    s      = seed
    inside = 0
    for _ in 1:n
        x, s = lcg_next(s)
        y, s = lcg_next(s)
        if x*x + y*y <= 1.0
            inside += 1
        end
    end
    return 4.0 * inside / n
end

function main()
    input = split(read(stdin, String))
    n     = parse(Int,    input[1])
    seed  = parse(UInt32, input[2])

    # JIT warmup
    montecarlo(100, UInt32(1))

    start = time_ns()
    pi_   = montecarlo(n, seed)
    ms    = (time_ns() - start) / 1e6

    println(@sprintf("%.8f", pi_))
    println(stderr, ms)
end

using Printf
main()

File: algorithms/python/montecarlo.py
import sys
import time

def lcg_next(s):
    s = (s * 1664525 + 1013904223) & 0xFFFFFFFF
    return (s >> 1) / (1 << 31), s

def montecarlo(n, seed):
    s      = seed
    inside = 0
    for _ in range(n):
        x, s = lcg_next(s)
        y, s = lcg_next(s)
        if x*x + y*y <= 1.0:
            inside += 1
    return 4.0 * inside / n

def main():
    data = sys.stdin.read().split()
    n    = int(data[0])
    seed = int(data[1])

    start = time.perf_counter()
    pi    = montecarlo(n, seed)
    ms    = (time.perf_counter() - start) * 1000

    sys.stdout.write(f"{pi:.8f}\n")
    sys.stderr.write(f"{ms:.3f}\n")

if __name__ == "__main__":
    main()